### PR TITLE
feat: Apollo-Server v3 and @graphql-tools update

### PR DESCRIPTION
**Pull request by viniciuslacerda0**
## Description

This PR aims to update `apollo-server` to v3 and `graphql-tools` to `@graphql-tools`.
- Since @graphql-tools/utils decided to transform `SchemaDirectiveVisitor` args in Generics, it is necessary to add some type assertion in some variables.

## Progress

- [x] apollo-server v3;
- [x] Use @graphql-tools instead of graphql-tools;

## Visual reference

image

### CODE REVIEW

File: lib/ValidateDirectiveVisitor.ts [hunk 1]
```diff
@@ -785,9 +785,11 @@ const collectInputObjectsAndFieldsWithArguments = (
  * will throw and the wrapped resolver won't be called.
  */
 abstract class ValidateDirectiveVisitor<
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   TArgs extends ValidationDirectiveArgs,
   TContext = object,
-> extends EasyDirectiveVisitor<TArgs> {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+> extends EasyDirectiveVisitor<any> {
```
Comment (ricardodalarme) [inline lib/ValidateDirectiveVisitor.ts:792]:
```suggestion
abstract class ValidateDirectiveVisitor<
  TArgs extends ValidationDirectiveArgs,
  TContext = object,
> extends EasyDirectiveVisitor<TArgs> {
```

File: lib/auth.ts [hunk 2]
```diff
@@ -20,6 +20,7 @@ export type AuthContext<TContext extends object = object> = {
 
 export class AuthDirectiveVisitor<
   TContext extends AuthContext,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
 > extends EasyDirectiveVisitor<{}> {
```
Comment (DDDKnightmare) [inline lib/auth.ts:24]: If there are no args, use  `Record<string, never>`
```suggestion
> extends EasyDirectiveVisitor<Record<string, never>, TContext> {
```

File: lib/EasyDirectiveVisitor.ts [hunk 3]
```diff
@@ -181,7 +181,7 @@ export const getDirectiveDeclaration = (
  */
 abstract class EasyDirectiveVisitor<
   TArgs extends object,
-> extends SchemaDirectiveVisitor {
+> extends SchemaDirectiveVisitor<TArgs> {
```
Comment (DDDKnightmare) [inline lib/EasyDirectiveVisitor.ts:184]: Add the context type

File: package.json [hunk 4]
```diff
@@ -69,6 +70,8 @@
     "graphql": "^14.7.0 || ^15.8.0"
   },
   "dependencies": {
+    "@graphql-tools/schema": "^7.1.5",
+    "@graphql-tools/utils": "^7.1.5",
     "graphql-tag": "^2.12.6",
     "graphql-tools": "^4.0.8",
```
Comment (ricardodalarme) [inline package.json:76]: Since you are now using every thing from @graphql-tools/schema and @graphql-tools/utils I think you can remove this package

File: package.json [hunk 5]
```diff
@@ -30,23 +30,24 @@
     "@types/express": "^4.17.7",
     "@types/serve-static": "^1.13.10",
     "apollo-server-caching": "^0.7.0",
-    "apollo-server-core": "^2.25.4",
+    "apollo-server-core": "^3.12.0",
     "apollo-server-env": "^3.1.0",
     "apollo-server-errors": "^2.5.0",
```
Comment (ricardodalarme) [inline package.json:35]: missing upgrading the apollo-server-errors

Comment (ricardodalarme) [inline package.json:35]: I think you can remove all these resolutions now

File: package.json [hunk 6]
```diff
@@ -69,6 +70,8 @@
     "graphql": "^14.7.0 || ^15.8.0"
```
Comment (ricardodalarme) [inline package.json:70]: To match the Apollo Server:

```suggestion
    "graphql": "^15.3.0 || ^16.0.0"
```

- https://www.npmjs.com/package/apollo-server?activeTab=code

File: package.json [hunk 7]
```diff
@@ -65,12 +65,13 @@
...
   },
   "peerDependencies": {
-    "apollo-server-errors": "^2.5.0",
+    "apollo-server-errors": "^3.3.1",
     "graphql": "^14.7.0 || ^15.8.0"
   },
   "dependencies": {
+    "@graphql-tools/schema": "^7.1.5",
+    "@graphql-tools/utils": "^7.1.5",
     "graphql-tag": "^2.12.6",
```
Comment (DDDKnightmare) [inline package.json:74]: nit: you removed graphql-tools from `package.json` 2 commits after removing all of its usages.

File: lib/ValidateDirectiveVisitor.ts [hunk 8]
```diff
@@ -786,8 +786,8 @@ const collectInputObjectsAndFieldsWithArguments = (
  */
 abstract class ValidateDirectiveVisitor<
   TArgs extends ValidationDirectiveArgs,
-  TContext = object,
-> extends EasyDirectiveVisitor<TArgs> {
+  TContext extends object,
+> extends EasyDirectiveVisitor<TArgs, TContext> {
```
Comment (barbieri) [inline lib/ValidateDirectiveVisitor.ts:790]: why did you remove the default?

File: examples/access-control-directives.ts [hunk 9]
```diff
@@ -58,7 +65,10 @@ const schema = makeExecutableSchema({
       throwIfMissingPermissions: (): number => 123,
     },
   },
-  schemaDirectives: { auth, hasPermissions },
+  schemaDirectives: schemaDirectives as Record<
+    string,
+    typeof SchemaDirectiveVisitor
```
Comment (barbieri) [inline examples/access-control-directives.ts:70]: I saw all of these casts... but to me it's not clear why this is needed.

Always put the tsc failure in the commit message (text, not screenshots ;-)). But please sync with me in our slack, so I can try to help

**Review by DDDKnightmare — APPROVED** [review]
